### DISCUSSION TITLE
The reduce filter may needs a flat suffix

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#7273] [Ideas]
In response to a question in talk: https://talk.tiddlywiki.org/t/how-to-get-the-sum-of-field-values-by-filtering-based-on-prefix-of-field-names/6199/28

It was asked how to calculate the sum of field values based on filed name for given tiddlers. @ericshulman provided a solution as below

```
\define makesum() [<currentTiddler>fields[]prefix[ct-]] :map[<..currentTiddler>get<currentTiddler>] +[sum[]]

$text text={{{ [tag[tidsToSum]] :map[subfilter<makesum>] +[sum[]] }}} />
```

Which works fine. BUT, I think this question lends itself to `reduce` operator. So the 

```
\define makesum() [<currentTiddler>fields[]prefix[ct-]] :reduce[<..currentTiddler>get<currentTiddler>add<accumulator>]

<$list filter="[tag[tidsToSum]] :reduce[subfilter<makesum>add<accumulator>]">

</$list>
```

Both solutions seem a little complex. BUT for `reduce`, if the input contains several values, it only takes the first value. For example, see below solution:

```

\define makesum() [<currentTiddler>fields[]prefix[ct-]] :map:flat[<..currentTiddler>get<currentTiddler>]

<$list filter="[tag[exmp195]] :reduce[subfilter<makesum>add<accumulator>]">

</$list>
```

The makesum here returns all values for all fields prefixed with `ct-`, BUT the `:reduce` only takes the first value and adds it to accumulator. 

What do you think if we add a `flat` suffix to `reduce`?
I think @saqimtiaz was the developer behind the reduce operator.

**Comment by pmario** [comment]
hmm, But that would mean, that the `:reduce` run-prefix would behave like `:map` ... 

:map and :reduce are already complex filter run prefixes. If they could be interchanged imo it would be even more complex. 

----

I would suggest, that we add the solution Eric provided to the docs. So there will be 1 way to solve that problem.

**Reply by kookma** [reply]
I think yes, we need such examples as the one by Eric to be added to doc. Understanding map and reduce needs more examples and I think both operators are quite powerful.

**Comment by ericshulman** [comment]
Here's another, more generic example of use the `:map` filter run prefix:
```
<!-- Get total size of all fields in a tiddler, including size of field NAMES as well as VALUES -->
\define size() [<currentTiddler>fields[]] :map[<..currentTiddler>get<currentTiddler>else[]length[]] [<currentTiddler>fields[]length[]] +[sum[]]

<!-- Get total size of all tiddlers -->
{{{ [all[tiddlers]] :map[subfilter<size>] +[sum[]] }}}
```
